translator/trace/jaeger/traces_to_jaegerproto.go

```diff
@@ -0,0 +1,361 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jaeger
+
+import (
+	"fmt"
+
+	"github.com/jaegertracing/jaeger/model"
+	otlptrace "github.com/open-telemetry/opentelemetry-proto/gen/go/trace/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+// InternalTracesToJaegerProto translates internal trace data into the Jaeger Proto for GRPC.
+// Returns slice of translated Jaeger batches and error if translation failed.
+func InternalTracesToJaegerProto(td pdata.Traces) ([]*model.Batch, error) {
+	resourceSpans := td.ResourceSpans()
+
+	if resourceSpans.Len() == 0 {
+		return nil, nil
+	}
+
+	batches := make([]*model.Batch, 0, resourceSpans.Len())
+
+	var translationErr error
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		if rs.IsNil() {
+			continue
+		}
+
+		batch, err := resourceSpansToJaegerProto(rs)
+		if err == nil && batch != nil {
+			batches = append(batches, batch)
+		} else {
+			translationErr = err
+		}
+	}
+
+	if len(batches) == 0 {
+		return nil, translationErr
+	}
+
+	// Do not report an error if any batches were translated
```

Comment:
Why? Do we want to silently ignore translation errors?

Reply:
Yes, this is what I'm not sure about. The same applied to here https://github.com/open-telemetry/opentelemetry-collector/pull/866/files#diff-d10598d4d76963cac94843e3a1b2dfd6R93 in order to avoid retries when some part of the batch were successfully sent out. 

Do you know any better way to approach it?

Reply:
In this particular case we could reject the whole batch if there is a translation error. It would be cleaner.

BTW I don't know how critical those errors are for jaeger, they are mostly rejecting invalid trace and span IDs. If it's possible to ignore those invalid IDs and send zeroes for example, that probably would be better. 

What do you think?

Reply:
I think it is safer to drop entire span and record a proper error (return Permanent error).

Reply:
changed